Integration File FileWritingMessageHandler do not close input stream, when the creation of output stream throws Security Exception

The `FileWritingMessageHandler`  do not close the `FileInputStream` opened on the source file if the creation on of the `FileOutputSource` on the destination faild for a `SecurityException`.

As a consequence you cannot remove the source file until the application stops. Even the application itself cannot remove the file in another thread.

I think the problem is in [Line 565 of FileWritingMessageHandler](https://github.com/spring-projects/spring-integration/blob/master/spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java#L565)

the instruction

			BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(tempFile), this.bufferSize);

should be inside the `try catch` block , otherwise if the constructor thows `java.lang.SecurityException`  the finally block that close the sourceFileInputStream is not called.

My work-around  soluction was to transform the payload from file to FileInputStream and override 

	protected Object handleRequestMessage(Message<?> requestMessage) 

in order to close the input stream if the `super.handleRequestMessage(Message<?> requestMessage)`
[FileReadingJavaApplication.java.txt](https://github.com/spring-projects/spring-integration/files/547589/FileReadingJavaApplication.java.txt)

  throws exception.


Step To Riproduce

- Create a simple spring boot application that use a FileReadingMessageSource as source and a FileWritingMessageHandler as Handler. No Trasformer between source and handler only a direct Channel. (see attachment).

- Remove the writing permission on the destination directory of  the FileWritingMessageHandler.

- Start the application and copy the a file  in the directory configured as source of FileReadingMessageSource .

- Correctly , the FileWritingMessageHandler  should failed and throw exception

`2016-10-24 13:05:38.390 ERROR 9796 --- [ask-scheduler-6] o.s.i.h.LoggingHandler                   : org.springframework.messaging.MessageHandlingException: failed to write Message payload to file; nested exception is java.io.FileNotFoundException: C:\SIRIO-SA\RecipeTo\localhost_access_log.2016-09-16.txt.writing (Accesso negato)`

- Do not stop the application and try to remove the file in the directory configured as source of FileReadingMessageSource . The file should be lock.



https://github.com/spring-projects/spring-integration/blob/master/spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java#L565

Yes, it could be a bug, but two observations from my side...

1. If the directory is not writable when the app starts, I get...

    10:18:21.066 [main] ERROR o.s.boot.SpringApplication - Application startup failed

    ...

    Caused by: java.lang.IllegalArgumentException: Destination directory [/tmp/bar] is not writable.

2. If I change the directory to be not writable after the app starts, I get...

    Caused by: java.lang.IllegalArgumentException: Destination directory [/tmp/bar] is not writable.
****

before we create the buffered input stream, or start writing; see the `canWrite()` test [here](https://github.com/spring-projects/spring-integration/blob/master/spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java#L397).

My tests are on Mac OS X.

Can you provide more information? i.e. Operating System, Spring Integration version, and a full stack trace.

I use Windows 10, with Spring spring-boot-starter-integration-1.4.1 .

I founded the bug by chance, I wanted to test how my application behaves when the copy failed. 

I recieve the runtime exeption at startup only if the directory do not exists

java.lang.IllegalArgumentException: Destination directory [C:\SIRIO-SA\RecipeTo] does not exist.

if it does exist but I just remove the write permissions (more precisely , I deny folder write to Everyone group on Windows) I don't get any runtime error at startup. only when the handler actually try to write the file I recieve the exception.

This is the complete exception stack trace

2016-10-24 16:33:46.762 ERROR 6292 --- [ask-scheduler-1] o.s.i.h.LoggingHandler                   : org.springframework.messaging.MessageHandlingException: failed to write Message payload to file; nested exception is java.io.FileNotFoundException: C:\SIRIO-SA\RecipeTo\localhost_access_log.2016-09-19.txt.writing (Accesso negato)
	at org.springframework.integration.file.FileWritingMessageHandler.handleRequestMessage(FileWritingMessageHandler.java:465)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:109)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:127)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:148)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:121)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:423)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:373)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:115)
	at org.springframework.messaging.core.GenericMessagingTemplate.doSend(GenericMessagingTemplate.java:45)
	at org.springframework.messaging.core.AbstractMessageSendingTemplate.send(AbstractMessageSendingTemplate.java:105)
	at org.springframework.integration.endpoint.SourcePollingChannelAdapter.handleMessage(SourcePollingChannelAdapter.java:195)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.doPoll(AbstractPollingEndpoint.java:272)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint.access$000(AbstractPollingEndpoint.java:58)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint$1.call(AbstractPollingEndpoint.java:190)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint$1.call(AbstractPollingEndpoint.java:186)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint$Poller$1.run(AbstractPollingEndpoint.java:353)
	at org.springframework.integration.util.ErrorHandlingTaskExecutor$1.run(ErrorHandlingTaskExecutor.java:55)
	at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50)
	at org.springframework.integration.util.ErrorHandlingTaskExecutor.execute(ErrorHandlingTaskExecutor.java:51)
	at org.springframework.integration.endpoint.AbstractPollingEndpoint$Poller.run(AbstractPollingEndpoint.java:344)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at org.springframework.scheduling.concurrent.ReschedulingRunnable.run(ReschedulingRunnable.java:81)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.FileNotFoundException: C:\SIRIO-SA\RecipeTo\localhost_access_log.2016-09-19.txt.writing (Accesso negato)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at org.springframework.integration.file.FileWritingMessageHandler.handleInputStreamMessage(FileWritingMessageHandler.java:566)
	at org.springframework.integration.file.FileWritingMessageHandler.handleFileMessage(FileWritingMessageHandler.java:510)
	at org.springframework.integration.file.FileWritingMessageHandler.handleRequestMessage(FileWritingMessageHandler.java:433)
	... 30 more

Confirmed on Windows 7:
```
org.springframework.messaging.MessageHandlingException: failed to write Message payload to file; nested exception is java.io.FileNotFoundException: target\base-directory\foo.txt.writing (Отказано в доступе)

	at org.springframework.integration.file.FileWritingMessageHandler.handleRequestMessage(FileWritingMessageHandler.java:464)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.handleMessageInternal(AbstractReplyProducingMessageHandler.java:109)
	at org.springframework.integration.handler.AbstractMessageHandler.handleMessage(AbstractMessageHandler.java:133)
	at org.springframework.integration.dispatcher.AbstractDispatcher.tryOptimizedDispatch(AbstractDispatcher.java:116)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.doDispatch(UnicastingDispatcher.java:148)
	at org.springframework.integration.dispatcher.UnicastingDispatcher.dispatch(UnicastingDispatcher.java:121)
	at org.springframework.integration.channel.AbstractSubscribableChannel.doSend(AbstractSubscribableChannel.java:77)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:423)
	at org.springframework.integration.channel.AbstractMessageChannel.send(AbstractMessageChannel.java:373)
	at org.springframework.integration.file.FileOutboundChannelAdapterIntegrationTests.saveToBaseDir(FileOutboundChannelAdapterIntegrationTests.java:98)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:73)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:83)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:250)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:189)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:262)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.io.FileNotFoundException: target\base-directory\foo.txt.writing (Отказано в доступе)
```
The:
```
Assert.isTrue(destinationDirectory.canWrite(),
				"Destination directory [" + destinationDirectory + "] is not writable.");
```
passes somehow. But at runtime we can't write because we fail already on the `InputStream` ctor:
```
BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(tempFile), this.bufferSize);
```

`Files.isWritable(destinationDirectory.toPath())` does the stuff for us though!
```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'org.springframework.integration.file.config.FileWritingMessageHandlerFactoryBean#0': FactoryBean threw exception on object creation; nested exception is org.springframework.beans.factory.BeanInitializationException: failed to initialize MessageHandler; nested exception is java.lang.IllegalArgumentException: Destination directory [target\base-directory] is not writable.
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:175)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.getObjectFromFactoryBean(FactoryBeanRegistrySupport.java:103)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getObjectForBeanInstance(AbstractBeanFactory.java:1598)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:252)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:195)
	at org.springframework.beans.factory.support.BeanDefinitionValueResolver.resolveReference(BeanDefinitionValueResolver.java:351)
	... 47 more
Caused by: org.springframework.beans.factory.BeanInitializationException: failed to initialize MessageHandler; nested exception is java.lang.IllegalArgumentException: Destination directory [target\base-directory] is not writable.
	at org.springframework.integration.config.AbstractSimpleMessageHandlerFactoryBean.createHandlerInternal(AbstractSimpleMessageHandlerFactoryBean.java:242)
	at org.springframework.integration.config.AbstractSimpleMessageHandlerFactoryBean.getObject(AbstractSimpleMessageHandlerFactoryBean.java:172)
	at org.springframework.integration.config.AbstractSimpleMessageHandlerFactoryBean.getObject(AbstractSimpleMessageHandlerFactoryBean.java:57)
	at org.springframework.beans.factory.support.FactoryBeanRegistrySupport.doGetObjectFromFactoryBean(FactoryBeanRegistrySupport.java:168)
	... 52 more
Caused by: java.lang.IllegalArgumentException: Destination directory [target\base-directory] is not writable.
	at org.springframework.util.Assert.isTrue(Assert.java:86)
	at org.springframework.integration.file.FileWritingMessageHandler.validateDestinationDirectory(FileWritingMessageHandler.java:397)
	at org.springframework.integration.file.FileWritingMessageHandler.doInit(FileWritingMessageHandler.java:351)
	at org.springframework.integration.handler.AbstractReplyProducingMessageHandler.onInit(AbstractReplyProducingMessageHandler.java:96)
	at org.springframework.integration.context.IntegrationObjectSupport.afterPropertiesSet(IntegrationObjectSupport.java:176)
	at org.springframework.integration.config.AbstractSimpleMessageHandlerFactoryBean.createHandlerInternal(AbstractSimpleMessageHandlerFactoryBean.java:239)
```

We can use Java 7 trick in the current `5.0` version and can try to hack reflection for the backport if Java 7 presents, but I guess the simple moving `new BufferedOutputStream(new FileOutputStream(tempFile), this.bufferSize);` will help us fully. Just because the permissions can be changed at runtime as well.

--- a/spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java
+++ b/spring-integration-file/src/main/java/org/springframework/integration/file/FileWritingMessageHandler.java
@@ -394,7 +394,7 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 				"Destination directory [" + destinationDirectory + "] does not exist.");
 		Assert.isTrue(destinationDirectory.isDirectory(),
 				"Destination path [" + destinationDirectory + "] does not point to a directory.");
-		Assert.isTrue(destinationDirectory.canWrite(),
+		Assert.isTrue(Files.isWritable(destinationDirectory.toPath()),
 				"Destination directory [" + destinationDirectory + "] is not writable.");
 	}
 
@@ -524,8 +524,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 
 				@Override
 				protected void whileLocked() throws IOException {
-					BufferedOutputStream bos = state != null ? state.stream : createOutputStream(fileToWriteTo, true);
+					BufferedOutputStream bos = null;
 					try {
+						bos = state != null ? state.stream : createOutputStream(fileToWriteTo, true);
 						byte[] buffer = new byte[StreamUtils.BUFFER_SIZE];
 						int bytesRead = -1;
 						while ((bytesRead = sourceFileInputStream.read(buffer)) != -1) {
@@ -543,7 +544,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 						}
 						try {
 							if (state == null || FileWritingMessageHandler.this.flushTask == null) {
-								bos.close();
+								if (bos != null) {
+									bos.close();
+								}
 								clearState(fileToWriteTo, state);
 							}
 							else {
@@ -562,9 +565,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 		}
 		else {
 
-			BufferedOutputStream bos = new BufferedOutputStream(new FileOutputStream(tempFile), this.bufferSize);
-
+			BufferedOutputStream bos = null;
 			try {
+				bos = new BufferedOutputStream(new FileOutputStream(tempFile), this.bufferSize);
 				byte[] buffer = new byte[StreamUtils.BUFFER_SIZE];
 				int bytesRead = -1;
 				while ((bytesRead = sourceFileInputStream.read(buffer)) != -1) {
@@ -582,7 +585,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 				catch (IOException ex) {
 				}
 				try {
-					bos.close();
+					if (bos != null) {
+						bos.close();
+					}
 				}
 				catch (IOException ex) {
 				}
@@ -605,8 +610,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 
 			@Override
 			protected void whileLocked() throws IOException {
-				BufferedOutputStream bos = state != null ? state.stream : createOutputStream(fileToWriteTo, append);
+				BufferedOutputStream bos = null;
 				try {
+					bos = state != null ? state.stream : createOutputStream(fileToWriteTo, append);
 					bos.write(bytes);
 					if (FileWritingMessageHandler.this.appendNewLine) {
 						bos.write(LINE_SEPARATOR.getBytes());
@@ -615,7 +621,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 				finally {
 					try {
 						if (state == null || FileWritingMessageHandler.this.flushTask == null) {
-							bos.close();
+							if (bos != null) {
+								bos.close();
+							}
 							clearState(fileToWriteTo, state);
 						}
 						else {
@@ -646,8 +654,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 
 			@Override
 			protected void whileLocked() throws IOException {
-				BufferedWriter writer = state != null ? state.writer : createWriter(fileToWriteTo, append);
+				BufferedWriter writer = null;
 				try {
+					writer = state != null ? state.writer : createWriter(fileToWriteTo, append);
 					writer.write(content);
 					if (FileWritingMessageHandler.this.appendNewLine) {
 						writer.newLine();
@@ -656,7 +665,9 @@ public class FileWritingMessageHandler extends AbstractReplyProducingMessageHand
 				finally {
 					try {
 						if (state == null || FileWritingMessageHandler.this.flushTask == null) {
-							writer.close();
+							if (writer != null) {
+								writer.close();
+							}
 							clearState(fileToWriteTo, state);
 						}
 						else {

